File: src/com/suushiemaniac/cubing/bld/model/enumeration/puzzle/TwistyPuzzle.java
package com.suushiemaniac.cubing.bld.model.enumeration.puzzle;

import com.suushiemaniac.cubing.alglib.alg.Algorithm;
import com.suushiemaniac.cubing.alglib.lang.NotationReader;
import com.suushiemaniac.cubing.bld.analyze.BldPuzzle;
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType;
import net.gnehzr.tnoodle.scrambles.Puzzle;

import java.util.function.Supplier;

public interface TwistyPuzzle {
	int getSize();

	Supplier<Puzzle> supplyScramblingPuzzle();

	BldPuzzle getAnalyzingPuzzle();

	PieceType[] getPieceTypes();

	NotationReader getReader();

	default Puzzle getScramblingPuzzle() {
		return this.supplyScramblingPuzzle().get();
	}

	default Algorithm generateRandomScramble() {
		return this.getReader().parse(this.getScramblingPuzzle().generateScramble());
	}

	default BldPuzzle getScrambleAnalysis(Algorithm scramble) {
		BldPuzzle analysis = this.getAnalyzingPuzzle();
		analysis.parseScramble(scramble);

		return analysis;
	}

	default BldPuzzle getScrambleAnalysis() {
		return this.getScrambleAnalysis(this.generateRandomScramble());
	}
}

File: src/main/kotlin/com/suushiemaniac/cubing/bld/model/enumeration/puzzle/TwistyPuzzle.kt
package com.suushiemaniac.cubing.bld.model.enumeration.puzzle

import com.suushiemaniac.cubing.alglib.alg.Algorithm
import com.suushiemaniac.cubing.alglib.lang.NotationReader
import com.suushiemaniac.cubing.bld.gsolve.KPuzzle
import com.suushiemaniac.cubing.bld.gsolve.GPuzzle
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType
import net.gnehzr.tnoodle.scrambles.Puzzle
import java.io.File

interface TwistyPuzzle {
    val size: Int

    val pieceTypes: Array<out PieceType>

    val kPuzzle: KPuzzle

    fun gPuzzle(gConfig: File) = this.kPuzzle.g(gConfig)

    val reader: NotationReader

    val scramblingPuzzle: Puzzle

    fun independentScrambleSupplier(): Sequence<Algorithm>

    val randomScramble: Algorithm
        get() = this.reader.parse(this.scramblingPuzzle.generateScramble())
}
